File: crates/prolly/bindings/ruby/examples/materialized_view.rb
# frozen_string_literal: true

require 'tmpdir'

$LOAD_PATH.unshift(File.expand_path('../lib', __dir__))
require 'prolly'

def upsert(key, value)
  Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: key.b, value: value.b)
end

def delete(key)
  Prolly::MutationRecord.new(kind: Prolly::MutationKind::DELETE, key: key.b, value: nil)
end

def assert_equal(expected, actual, label = 'value')
  return if expected == actual

  raise "#{label}: expected #{expected.inspect}, got #{actual.inspect}"
end

Order = Struct.new(:tenant, :id, :status, :cents)

def order_key(order)
  "orders/source/tenant/#{order.tenant}/order/#{order.id}".b
end

def encode_order(order)
  "#{order.tenant}|#{order.id}|#{order.status}|#{order.cents}".b
end

def decode_order(value)
  tenant, id, status, cents = value.split('|'.b)
  Order.new(tenant, id, status, cents.to_i)
end

def view_key(tenant, status)
  "orders/view/by-status/tenant/#{tenant}/status/#{status}".b
end

def build_revenue_view(engine, source)
  totals = Hash.new(0)
  engine.range(source, 'orders/source/'.b, 'orders/source0'.b).each do |entry|
    order = decode_order(entry.value)
    totals[[order.tenant, order.status]] += order.cents
  end
  mutations = totals.sort.map do |(tenant, status), cents|
    Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: view_key(tenant, status), value: cents.to_s.b)
  end
  engine.batch(engine.create, mutations)
end

def materialized_view
  engine = Prolly::ProllyEngine.memory(Prolly.default_config)
  o1 = Order.new('acme', 'o1', 'paid', 1200)
  o2 = Order.new('acme', 'o2', 'open', 500)
  source_v1 = engine.batch(
    engine.create,
    [
      Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: order_key(o1), value: encode_order(o1)),
      Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: order_key(o2), value: encode_order(o2))
    ]
  )
  paid_o2 = Order.new('acme', 'o2', 'paid', 500)
  source_v2 = engine.put(source_v1, order_key(paid_o2), encode_order(paid_o2))
  view_v2 = build_revenue_view(engine, source_v2)

  assert_equal '1700'.b, engine.get(view_v2, view_key('acme', 'paid')), 'paid revenue'
  assert_equal nil, engine.get(view_v2, view_key('acme', 'open')), 'open revenue'

  puts "materialized_view: folded #{engine.diff(source_v1, source_v2).length} source diff"
end

materialized_view

File: crates/prolly/bindings/python/examples/materialized_view.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from tempfile import TemporaryDirectory
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import prolly


def b(value: str) -> bytes:
    return value.encode("utf-8")


def text(value: bytes | None) -> str | None:
    return None if value is None else value.decode("utf-8")


def assert_bytes(expected: bytes, actual: bytes | None, label: str) -> None:
    if actual != expected:
        raise AssertionError(f"{label}: expected {expected!r}, got {actual!r}")


def upsert(key: str, value: str | bytes) -> prolly.MutationRecord:
    payload = value if isinstance(value, bytes) else b(value)
    return prolly.MutationRecord(kind=prolly.MutationKind.UPSERT, key=b(key), value=payload)


def delete(key: str) -> prolly.MutationRecord:
    return prolly.MutationRecord(kind=prolly.MutationKind.DELETE, key=b(key), value=None)

@dataclass(frozen=True)
class Order:
    tenant: str
    order_id: str
    status: str
    cents: int


def encode_order(order: Order) -> bytes:
    return b(f"{order.tenant}|{order.order_id}|{order.status}|{order.cents}")


def decode_order(value: bytes) -> Order:
    tenant, order_id, status, cents = text(value).split("|")
    return Order(tenant, order_id, status, int(cents))


def order_key(order: Order) -> bytes:
    return b(f"orders/source/tenant/{order.tenant}/order/{order.order_id}")


def view_key(tenant: str, status: str) -> bytes:
    return b(f"orders/view/by-status/tenant/{tenant}/status/{status}")


def build_revenue_view(engine: prolly.ProllyEngine, source: prolly.TreeRecord) -> prolly.TreeRecord:
    totals: dict[tuple[str, str], int] = {}
    for entry in engine.range(source, b"orders/source/", b"orders/source0"):
        order = decode_order(entry.value)
        totals[(order.tenant, order.status)] = totals.get((order.tenant, order.status), 0) + order.cents

    mutations = [
        prolly.MutationRecord(kind=prolly.MutationKind.UPSERT, key=view_key(tenant, status), value=str(cents).encode())
        for (tenant, status), cents in sorted(totals.items())
    ]
    return engine.batch(engine.create(), mutations)

def materialized_view() -> None:
    engine = prolly.ProllyEngine.memory(prolly.default_config())
    source_v1 = engine.batch(
        engine.create(),
        [
            prolly.MutationRecord(kind=prolly.MutationKind.UPSERT, key=order_key(Order("acme", "o1", "paid", 1200)), value=encode_order(Order("acme", "o1", "paid", 1200))),
            prolly.MutationRecord(kind=prolly.MutationKind.UPSERT, key=order_key(Order("acme", "o2", "open", 500)), value=encode_order(Order("acme", "o2", "open", 500))),
        ],
    )
    source_v2 = engine.put(source_v1, order_key(Order("acme", "o2", "paid", 500)), encode_order(Order("acme", "o2", "paid", 500)))
    view_v2 = build_revenue_view(engine, source_v2)

    assert_bytes(b"1700", engine.get(view_v2, view_key("acme", "paid")), "paid revenue")
    assert engine.get(view_v2, view_key("acme", "open")) is None

    print(f"materialized_view: folded {len(engine.diff(source_v1, source_v2))} source diff")


if __name__ == "__main__":
    materialized_view()
